File: shit/frontend/src/hooks/Layouts.ts
/**
 * src/hooks/useLayouts.ts
 */

import { useState, useEffect, useCallback } from 'react';

export interface CustomCell {
    id: string;
    row: number;
    col: number;
    rowSpan: number;
    colSpan: number;
}

export interface SavedLayout {
    name: string;
    gridSize: number | 'custom' | 'single';
    customCells?: CustomCell[];
    customGridRows?: number;
    customGridCols?: number;
    activeCells: Record<string, string>;
    // Состояние вывода 360 на момент сохранения; киоск применяет при загрузке
    surround?: { viewMode: 'top' | 'surround'; manual: boolean };
    timestamp: number;
}

// Человекочитаемое сообщение по статусу ответа
function httpMsg(status: number, action: string): string {
    if (status === 404) return `Эндпоинт ${action} не найден (404) — роутер layouts не подключён?`;
    if (status === 422) return `Ошибка валидации данных (422)`;
    if (status === 503) return `Сервер недоступен (503)`;
    return `Ошибка ${status} при ${action}`;
}

// ── API helpers — никогда не бросают, возвращают { data | null, error | null } ──

async function apiGetLayouts(): Promise<{ data: SavedLayout[] | null; error: string | null }> {
    try {
        const res = await fetch('/api/layouts');
        if (!res.ok) return { data: null, error: httpMsg(res.status, 'загрузке сеток') };
        const data = await res.json();
        return { data: Array.isArray(data) ? data : [], error: null };
    } catch {
        return { data: null, error: 'Нет связи с сервером при загрузке сеток' };
    }
}

async function apiUpsertLayout(layout: SavedLayout): Promise<{ data: SavedLayout | null; error: string | null }> {
    try {
        const res = await fetch('/api/layouts', {
            method:  'POST',
            headers: { 'Content-Type': 'application/json' },
            body:    JSON.stringify(layout),
        });
        if (!res.ok) return { data: null, error: httpMsg(res.status, 'сохранении сетки') };
        return { data: await res.json(), error: null };
    } catch {
        return { data: null, error: 'Не удалось сохранить сетку — нет связи с сервером' };
    }
}

async function apiDeleteLayout(name: string): Promise<{ error: string | null }> {
    try {
        const res = await fetch(`/api/layouts/${encodeURIComponent(name)}`, { method: 'DELETE' });
        if (!res.ok && res.status !== 404)
            return { error: httpMsg(res.status, 'удалении сетки') };
        return { error: null };
    } catch {
        return { error: 'Не удалось удалить сетку — нет связи с сервером' };
    }
}

// ── Hook ──────────────────────────────────────────────────────

export function useLayouts() {
    const [layouts, setLayouts] = useState<SavedLayout[]>([]);
    const [loading, setLoading] = useState(true);
    // loadError — ошибка начальной загрузки (показывается в UI как предупреждение)
    const [loadError, setLoadError] = useState<string>('');
    // opError — ошибка save/remove (показывается кратко рядом с действием)
    const [opError, setOpError] = useState<string>('');

    const load = useCallback(async () => {
        setLoading(true);
        setLoadError('');
        const { data, error } = await apiGetLayouts();
        if (error) {
            setLoadError(error);
            setLayouts([]);          // пустой список — не крашимся
        } else {
            setLayouts(data ?? []);
        }
        setLoading(false);
    }, []);

    useEffect(() => { load(); }, [load]);

    const save = useCallback(async (layout: SavedLayout): Promise<boolean> => {
        setOpError('');
        const { data, error } = await apiUpsertLayout(layout);
        if (error) { setOpError(error); return false; }
        if (!data) return false;
        setLayouts(prev => {
            const idx = prev.findIndex(l => l.name === data.name);
            if (idx >= 0) { const next = [...prev]; next[idx] = data; return next; }
            return [...prev, data];
        });
        return true;
    }, []);

    const remove = useCallback(async (name: string): Promise<boolean> => {
        setOpError('');
        const { error } = await apiDeleteLayout(name);
        if (error) { setOpError(error); return false; }
        setLayouts(prev => prev.filter(l => l.name !== name));
        return true;
    }, []);

    const clearOpError = useCallback(() => setOpError(''), []);

    return { layouts, loading, loadError, opError, clearOpError, reload: load, save, remove };
}

File: shit/backend/app/routers/layouts.py
"""
app/routers/layouts.py

Хранение видеосеток (layouts) на устройстве в JSON-файле.
Маршруты:
  GET    /api/layouts          — список всех сеток
  POST   /api/layouts          — создать/обновить сетку (по name как ключу)
  DELETE /api/layouts/{name}   — удалить сетку по имени
"""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, Dict, List

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel

from app.config import settings

router = APIRouter()
logger = logging.getLogger(__name__)

# Том настроек мастера: переживает пересборку контейнеров
LAYOUTS_FILE = Path(settings.LAYOUTS_FILE)


# ── Pydantic-модели ────────────────────────────────────────────

class CustomCell(BaseModel):
    id: str
    row: int
    col: int
    rowSpan: int
    colSpan: int


class Layout(BaseModel):
    name: str
    gridSize: Any                          # 1 | 4 | 9 | 16 | "custom" | "single"
    customCells: List[CustomCell] | None = None
    customGridRows: int | None = None
    customGridCols: int | None = None
    activeCells: Dict[str, str]            # { cellId: cameraId }
    # Состояние вывода 360: { viewMode: "top"|"surround", manual: bool }
    surround: Dict[str, Any] | None = None
    # Камеры birdview с включённой коррекцией дисторсии: { cameraId: true }
    corrections: Dict[str, bool] | None = None
    timestamp: int


# ── Helpers ────────────────────────────────────────────────────

def _read_layouts() -> List[Dict]:
    """Читает layouts.json, возвращает список dict. Если файла нет — пустой список."""
    try:
        if LAYOUTS_FILE.exists():
            return json.loads(LAYOUTS_FILE.read_text(encoding="utf-8"))
    except Exception as e:
        logger.error(f"layouts: read error: {e}")
    return []


def _write_layouts(layouts: List[Dict]) -> None:
    """Записывает список dict в layouts.json атомарно."""
    try:
        LAYOUTS_FILE.parent.mkdir(parents=True, exist_ok=True)
        tmp = LAYOUTS_FILE.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(layouts, ensure_ascii=False, indent=2), encoding="utf-8")
        tmp.replace(LAYOUTS_FILE)
    except Exception as e:
        logger.error(f"layouts: write error: {e}")
        raise


# ── Endpoints ──────────────────────────────────────────────────

@router.get("/layouts", response_model=List[Layout])
async def get_layouts():
    """Вернуть все сохранённые сетки."""
    return _read_layouts()


@router.post("/layouts", response_model=Layout, status_code=status.HTTP_200_OK)
async def upsert_layout(layout: Layout):
    """
    Создать новую сетку или обновить существующую (ключ — name).
    Timestamp ставит клиент.
    """
    layouts = _read_layouts()

    idx = next((i for i, l in enumerate(layouts) if l.get("name") == layout.name), None)
    data = layout.dict()

    if idx is not None:
        layouts[idx] = data
    else:
        layouts.append(data)

    _write_layouts(layouts)
    return data


@router.delete("/layouts/{name}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_layout(name: str):
    """Удалить сетку по имени."""
    layouts = _read_layouts()
    filtered = [l for l in layouts if l.get("name") != name]

    if len(filtered) == len(layouts):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Layout '{name}' not found",
        )

    _write_layouts(filtered)
    return None